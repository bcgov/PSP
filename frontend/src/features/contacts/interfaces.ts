export interface IContactFilter {
  summary: string;
  municipality: string;
  activeContactsOnly: boolean;
  searchBy: string;
}

export enum ContactTypes {
  ORGANIZATION = 'O',
  INDIVIDUAL = 'P',
}
