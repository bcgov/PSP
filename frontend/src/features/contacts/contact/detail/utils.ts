import { IContactAddress } from 'interfaces/IContact';

import { AddressTypes } from './../../../../constants/addressTypes';

// the order of this array corresponds to the expected display order
const addressSortOrder = [AddressTypes.Mailing, AddressTypes.Residential, AddressTypes.Billing];

export const sortAddresses = (a1: IContactAddress, a2: IContactAddress) => {
  if (a2.addressType.id === a1.addressType.id) {
    return 0;
  }
  const a2Index = addressSortOrder.indexOf(a2.addressType.id as AddressTypes);
  if (a2Index !== -1 && a2Index < addressSortOrder.indexOf(a1.addressType.id as AddressTypes)) {
    return 1;
  }

  return -1;
};

export const fakeAddresses = [
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'BILLING',
      description: 'Billing address',
      isDisabled: false,
    },
    streetAddress1: 'Billing address',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6A 5G7',
  },
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'MAILING',
      description: 'Mailing address',
      isDisabled: false,
    },
    streetAddress1: 'Mailing address',
    streetAddress2: 'Living in a van',
    streetAddress3: 'Down by the River',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6Z 5G7',
  },
  {
    id: 5,
    rowVersion: 1,
    addressType: {
      id: 'RESIDNT',
      description: 'Property address',
      isDisabled: false,
    },
    streetAddress1: 'Property address',
    streetAddress2: '',
    streetAddress3: '',
    municipality: 'vic',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V9A 0H6',
  },
];
