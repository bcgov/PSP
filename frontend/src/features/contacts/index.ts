export * from './contact/breadcrumb/ContactBreadcrumb';
export * from './contact/detail/Container';
export * from './contact/typeSelector/ContactTypeSelector';
export * from './interfaces';
export * from './list/ContactListView';
export * from './styles';
