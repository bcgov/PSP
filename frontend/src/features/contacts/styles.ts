import { Row } from 'react-bootstrap';
import styled from 'styled-components';

export const H1 = styled.h1`
  text-align: left;
`;

export const H2 = styled.h2`
  text-align: left;
`;

export const H2Primary = styled.h2`
  text-align: left;
  color: ${props => props.theme.css.primaryColor};
`;

export const RowAligned = styled(Row)`
  text-align: left;
`;

export const StatusIndicators = styled.div`
  border-radius: 1rem;
  background-color: white;
  border: 1px solid ${props => props.theme.css.lightVariantColor};
  padding: 0.2rem 1rem;
  color: ${props => props.theme.css.lightVariantColor};
  &.active {
    color: ${props => props.theme.css.completedColor};
    border: 1px solid ${props => props.theme.css.completedColor};
  }
`;
