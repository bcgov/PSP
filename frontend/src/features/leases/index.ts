export * from './add/AddLeaseContainer';
export * from './add/AddLeaseContainer';
export * from './add/AddLeaseLayout';
export * from './add/PropertyRow';
export * from './detail/BackToSearchButton';
export * from './detail/LeaseBreadCrumb';
export * from './detail/LeaseContainer';
export * from './detail/LeaseHeader';
export * from './detail/LeaseIndex';
export * from './detail/LeaseLayout';
export * from './detail/LeasePages/details/DetailAdministration';
export * from './detail/LeasePages/details/DetailDescription';
export * from './detail/LeasePages/details/DetailNotes';
export * from './detail/LeasePages/details/Details';
export * from './detail/LeasePages/details/DetailTermInformation';
export * from './detail/LeasePages/details/DetailTermInformationBox';
export * from './detail/LeasePages/details/DetailTerms';
export * from './detail/LeasePages/details/PropertiesInformation';
export * from './detail/LeasePages/details/PropertyInformation';
export * from './detail/LeasePages/LeasePageForm';
export * from './detail/LeaseRouter';
export * from './detail/LeaseStatusSummary';
export * from './detail/StackedPidTenantFields';
export * from './hooks/useLeaseDetail';
export * from './interfaces';
export * from './list/LeaseListView';
