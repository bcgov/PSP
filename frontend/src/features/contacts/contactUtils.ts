import { emailContactMethods, phoneContactMethods } from 'constants/contactMethodType';
import { AddressTypes } from 'constants/index';
import {
  getDefaultAddress,
  getDefaultContactMethod,
  ICreateOrganization,
  ICreateOrganizationForm,
  IEditableContactMethod,
  IEditableContactMethodForm,
  IEditablePerson,
  IEditablePersonAddress,
  IEditablePersonAddressForm,
  IEditablePersonForm,
} from 'interfaces/editable-contact';
import { stringToBoolean, stringToNull, stringToTypeCode, typeCodeToString } from 'utils/formUtils';

export function formPersonToApiPerson(formValues: IEditablePersonForm): IEditablePerson {
  // exclude form-specific fields from API payload object
  const {
    mailingAddress,
    propertyAddress,
    billingAddress,
    emailContactMethods,
    phoneContactMethods,
    ...restObject
  } = formValues;

  const addresses = [mailingAddress, propertyAddress, billingAddress]
    .filter(hasAddress)
    .map(formAddressToApiAddress);

  const contactMethods = [...emailContactMethods, ...phoneContactMethods]
    .filter(hasContactMethod)
    .map(formContactMethodToApiContactMethod);

  const apiPerson = {
    ...restObject,
    isDisabled: stringToBoolean(formValues.isDisabled),
    addresses,
    contactMethods,
  } as IEditablePerson;

  return apiPerson;
}

export function apiPersonToFormPerson(person?: IEditablePerson) {
  if (!person) return undefined;

  // exclude api-specific fields from form values
  const { addresses, contactMethods, ...restObject } = person;

  // split address array into sub-types: MAILING, RESIDENTIAL, BILLING
  const formAddresses = addresses?.map(apiAddressToFormAddress) || [];
  const addressDictionary = toDictionary(formAddresses, 'addressTypeId');

  // split contact methods array into phone and email values
  const formContactMethods = contactMethods?.map(apiContactMethodToFormContactMethod) || [];
  const emailContactMethods = formContactMethods.filter(isEmail);
  const phoneContactMethods = formContactMethods.filter(isPhone);

  const formPerson = {
    ...restObject,
    mailingAddress:
      addressDictionary[AddressTypes.Mailing] ?? getDefaultAddress(AddressTypes.Mailing),
    propertyAddress:
      addressDictionary[AddressTypes.Residential] ?? getDefaultAddress(AddressTypes.Residential),
    billingAddress:
      addressDictionary[AddressTypes.Billing] ?? getDefaultAddress(AddressTypes.Billing),
    emailContactMethods:
      emailContactMethods.length > 0 ? emailContactMethods : [getDefaultContactMethod()],
    phoneContactMethods:
      phoneContactMethods.length > 0 ? phoneContactMethods : [getDefaultContactMethod()],
  } as IEditablePersonForm;

  return formPerson;
}

export function organizationCreateFormToApiOrganization(
  formValues: ICreateOrganizationForm,
): ICreateOrganization {
  // exclude form-specific fields from API payload object
  const {
    mailingAddress,
    propertyAddress,
    billingAddress,
    emailContactMethods,
    phoneContactMethods,
    ...restObject
  } = formValues;

  const addresses = [mailingAddress, propertyAddress, billingAddress]
    .filter(hasAddress)
    .map(formAddressToApiAddress);

  const contactMethods = [...emailContactMethods, ...phoneContactMethods]
    .filter(hasContactMethod)
    .map(formContactMethodToApiContactMethod);

  const apiOrganization = {
    ...restObject,
    addresses,
    contactMethods,
  } as ICreateOrganization;

  return apiOrganization;
}

function hasContactMethod(formContactMethod?: IEditableContactMethodForm): boolean {
  if (!formContactMethod) return false;

  const { value, contactMethodTypeCode } = formContactMethod;
  return value !== '' && contactMethodTypeCode !== '';
}

function hasAddress(formAddress?: IEditablePersonAddressForm): boolean {
  if (!formAddress) return false;

  let { streetAddress1, addressTypeId, countryId, provinceId, municipality, postal } = formAddress;
  countryId = parseInt(countryId.toString()) || 0;
  provinceId = parseInt(provinceId.toString()) || 0;

  return (
    streetAddress1 !== '' &&
    addressTypeId !== '' &&
    municipality !== '' &&
    postal !== '' &&
    countryId > 0 &&
    provinceId > 0
  );
}

function formAddressToApiAddress(formAddress: IEditablePersonAddressForm): IEditablePersonAddress {
  return {
    ...formAddress,
    countryId: parseInt(formAddress.countryId.toString()) || 0,
    provinceId: parseInt(formAddress.provinceId.toString()) || 0,
    addressTypeId: stringToTypeCode(formAddress.addressTypeId),
  } as IEditablePersonAddress;
}

function apiAddressToFormAddress(address?: IEditablePersonAddress) {
  if (!address) return undefined;

  return {
    ...address,
    addressTypeId: typeCodeToString(address.addressTypeId),
  } as IEditablePersonAddressForm;
}

function formContactMethodToApiContactMethod(formContactMethod: IEditableContactMethodForm) {
  return {
    ...formContactMethod,
    value: stringToNull(formContactMethod.value),
    contactMethodTypeCode: stringToTypeCode(formContactMethod.contactMethodTypeCode),
  } as IEditableContactMethod;
}

function apiContactMethodToFormContactMethod(contactMethod?: IEditableContactMethod) {
  if (!contactMethod) return undefined;

  return {
    ...contactMethod,
    contactMethodTypeCode: typeCodeToString(contactMethod.contactMethodTypeCode),
  } as IEditableContactMethodForm;
}

// utility function to map an array to a object dictionary based on a given key
// used to convert the array of addresses into 3 separate fields: mailingAddress, propertyAddress and billingAddress
function toDictionary(array: any[], key: string) {
  return Object.assign({}, ...array.map(obj => ({ [obj[key]]: obj })));
}

function isEmail(contactMethod?: IEditableContactMethodForm): boolean {
  return !!contactMethod && emailContactMethods.includes(contactMethod.contactMethodTypeCode);
}

function isPhone(contactMethod?: IEditableContactMethodForm): boolean {
  return !!contactMethod && phoneContactMethods.includes(contactMethod.contactMethodTypeCode);
}
