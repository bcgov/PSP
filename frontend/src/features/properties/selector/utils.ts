import { IMapProperty } from 'features/properties/selector/models';
import { compact } from 'lodash';
import { Api_PropertyFile } from 'models/api/PropertyFile';
import { pidFormatter } from 'utils';

export enum NameSourceType {
  PID = 'PID',
  PIN = 'PIN',
  PLAN = 'Plan #',
  LOCATION = 'Location',
  NAME = 'Descriptive Name',
  NONE = 'n/a',
}

interface PropertyName {
  label: NameSourceType;
  value: string;
}

export const getPropertyName = (property: IMapProperty): PropertyName => {
  if (property.pid !== undefined && property.pid?.toString().length > 0 && property.pid !== '0') {
    return { label: NameSourceType.PID, value: pidFormatter(property.pid.toString()) };
  } else if (
    property.pin !== undefined &&
    property.pin?.toString()?.length > 0 &&
    property.pin !== '0'
  ) {
    return { label: NameSourceType.PIN, value: property.pin.toString() };
  } else if (property.planNumber !== undefined) {
    return { label: NameSourceType.PLAN, value: property.planNumber };
  } else if (property.latitude !== undefined && property.longitude !== undefined) {
    return {
      label: NameSourceType.LOCATION,
      value: compact([property.latitude?.toFixed(5), property.longitude?.toFixed(5)]).join(', '),
    };
  }
  return { label: NameSourceType.NONE, value: '' };
};

export const getFilePropertyName = (
  fileProperty?: Api_PropertyFile,
  skipName: boolean = false,
): PropertyName => {
  if (fileProperty === undefined) {
    return { label: NameSourceType.NONE, value: '' };
  }

  if (
    fileProperty.propertyName !== undefined &&
    fileProperty.propertyName !== '' &&
    skipName === false
  ) {
    return { label: NameSourceType.NAME, value: fileProperty.propertyName };
  } else if (fileProperty.property !== undefined) {
    const property = fileProperty.property;
    let mapProperty: IMapProperty = {
      pin: property.pin?.toString(),
      pid: property.pid?.toString(),
      latitude: property.latitude,
      longitude: property.longitude,
      planNumber: property.planNumber,
    };
    return getPropertyName(mapProperty);
  }
  return { label: NameSourceType.NONE, value: '' };
};
