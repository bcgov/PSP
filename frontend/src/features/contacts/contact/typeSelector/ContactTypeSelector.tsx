import { InlineFlexDiv } from 'components/common/styles';
import { ContactTypes } from 'features/contacts';
import * as React from 'react';
import { Form } from 'react-bootstrap';
import styled from 'styled-components';

interface IContactTypeSelectorProps {
  disabled?: boolean;
  contactType: ContactTypes;
  setContactType: (contactType: ContactTypes) => void;
}

export const ContactTypeSelector: React.FunctionComponent<IContactTypeSelectorProps> = ({
  contactType,
  setContactType,
  disabled,
}) => {
  return (
    <StyledInlineFlex>
      <Form.Check
        id="contact-individual"
        type="radio"
        label="Individual"
        disabled={disabled}
        checked={contactType === ContactTypes.INDIVIDUAL}
        onChange={(e: React.SyntheticEvent) => setContactType(ContactTypes.INDIVIDUAL)}
      ></Form.Check>
      <Form.Check
        id="contact-organization"
        label="Organization"
        type="radio"
        disabled={disabled}
        checked={contactType === ContactTypes.ORGANIZATION}
        onChange={(e: React.SyntheticEvent) => setContactType(ContactTypes.ORGANIZATION)}
      ></Form.Check>
    </StyledInlineFlex>
  );
};

const StyledInlineFlex = styled(InlineFlexDiv)`
  gap: 5rem;
  background-color: ${props => props.theme.css.filterBackgroundColor};
  padding: 0.5rem 3rem;
  border-radius: 0.5rem;
  margin-right: 2.5rem;
  .form-check {
    display: flex;
    align-items: center;
    padding-left: 2rem;
  }
`;

export default ContactTypeSelector;
