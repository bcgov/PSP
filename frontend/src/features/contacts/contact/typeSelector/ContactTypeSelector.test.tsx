import userEvent from '@testing-library/user-event';
import { ContactTypes } from 'features/contacts';
import { createMemoryHistory } from 'history';
import { render, RenderOptions } from 'utils/test-utils';

import { ContactTypeSelector } from './ContactTypeSelector';

const history = createMemoryHistory();
const setContactTypeFn = jest.fn();

describe('ContactTypeSelector component', () => {
  const setup = (
    renderOptions: RenderOptions & {
      contactType?: ContactTypes;
      setContactType?: (contactType: ContactTypes) => void;
    } = {},
  ) => {
    // render component under test
    const component = render(
      <ContactTypeSelector
        contactType={renderOptions.contactType ?? ContactTypes.INDIVIDUAL}
        setContactType={renderOptions.setContactType ?? setContactTypeFn}
      />,
      {
        ...renderOptions,
        history,
      },
    );

    return {
      component,
    };
  };
  beforeEach(() => {
    setContactTypeFn.mockReset();
  });
  it('renders as expected', () => {
    const { component } = setup();
    expect(component.asFragment()).toMatchSnapshot();
  });
  it('checks individual by default', () => {
    const {
      component: { getByLabelText },
    } = setup();
    expect(getByLabelText('Individual')).toBeChecked();
  });
  it('highlights individual if selected', () => {
    const {
      component: { getByLabelText },
    } = setup({ contactType: ContactTypes.INDIVIDUAL });
    expect(getByLabelText('Individual')).toBeChecked();
  });
  it('highlights organization if selected', () => {
    const {
      component: { getByLabelText },
    } = setup({ contactType: ContactTypes.ORGANIZATION });
    expect(getByLabelText('Organization')).toBeChecked();
  });
  it('calls setContactType correctly if individual is clicked', () => {
    const {
      component: { getByLabelText },
    } = setup({ contactType: ContactTypes.ORGANIZATION });
    userEvent.click(getByLabelText('Individual'));
    expect(setContactTypeFn).toHaveBeenCalledWith(ContactTypes.INDIVIDUAL);
  });

  it('calls setContactType correctly if organization is clicked', () => {
    const {
      component: { getByLabelText },
    } = setup({ contactType: ContactTypes.INDIVIDUAL });
    userEvent.click(getByLabelText('Organization'));
    expect(setContactTypeFn).toHaveBeenCalledWith(ContactTypes.ORGANIZATION);
  });
});
