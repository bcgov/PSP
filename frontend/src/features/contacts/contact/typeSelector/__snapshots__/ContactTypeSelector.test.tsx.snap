// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactTypeSelector component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="Toastify"
  />
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.c1 {
  gap: 5rem;
  padding: 0.5rem 3rem;
  border-radius: 0.5rem;
  margin-right: 2.5rem;
}

.c1 .form-check {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 2rem;
}

<div
    class="c0 c1"
  >
    <div
      class="form-check"
    >
      <input
        checked=""
        class="form-check-input"
        id="contact-individual"
        type="radio"
      />
      <label
        class="form-check-label"
        for="contact-individual"
        title=""
      >
        Individual
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="contact-organization"
        type="radio"
      />
      <label
        class="form-check-label"
        for="contact-organization"
        title=""
      >
        Organization
      </label>
    </div>
  </div>
</DocumentFragment>
`;
